llvm/lib/Target/AVR/MCTargetDesc/AVRAsmBackend.cpp

```diff
@@ -31,60 +31,72 @@ namespace adjust {
 
 using namespace llvm;
 
-static void signed_width(unsigned Width, uint64_t Value,
-                         std::string Description, const MCFixup &Fixup,
-                         MCContext *Ctx) {
-  if (!isIntN(Width, Value)) {
-    std::string Diagnostic = "out of range " + Description;
+static bool checkSignedWidth(unsigned Width, uint64_t Value,
+                             std::string Description, const MCFixup &Fixup,
+                             MCContext *Ctx) {
+  if (isIntN(Width, Value)) {
+    return true;
+  }
 
+  if (Ctx) {
     int64_t Min = minIntN(Width);
     int64_t Max = maxIntN(Width);
 
-    Diagnostic += " (expected an integer in the range " + std::to_string(Min) +
-                  " to " + std::to_string(Max) + ")";
-
-    Ctx->reportError(Fixup.getLoc(), Diagnostic);
+    Ctx->reportError(Fixup.getLoc(), Twine("out of range ") + Description +
+                                         " (expected an integer in the range " +
+                                         Twine(Min) + " to " + Twine(Max) +
+                                         ")");
   }
+
+  return false;
 }
 
-static void unsigned_width(unsigned Width, uint64_t Value,
-                           std::string Description, const MCFixup &Fixup,
-                           MCContext *Ctx) {
-  if (!isUIntN(Width, Value)) {
-    std::string Diagnostic = "out of range " + Description;
+static bool checkUnsignedWidth(unsigned Width, uint64_t Value,
+                               std::string Description, const MCFixup &Fixup,
+                               MCContext *Ctx) {
+  if (isUIntN(Width, Value)) {
+    return true;
+  }
 
+  if (Ctx) {
     int64_t Max = maxUIntN(Width);
 
-    Diagnostic +=
-        " (expected an integer in the range 0 to " + std::to_string(Max) + ")";
-
-    Ctx->reportError(Fixup.getLoc(), Diagnostic);
+    Ctx->reportError(Fixup.getLoc(),
```

Comment:
Missing error test

Reply:
Any idea how to test it?

This is called mostly for `adjust_call()`, but since that instruction takes a 22-bit offset, we can't use the same test-trick with repeated `call void asm sideeffect "nop", ""()` that we use for `branch-relaxation*.ll`.

Reply:
You can write an assembler test and there are various directives you can use to get a lot of binary padding, like .align and .fill

Reply:
Unfortunately it seems that the `Value` there is always zero, for all types of fixups (for calls, `fixup_port6` etc., you name it); can't wrap my head around why that's happening 🤔

Reply:
Unfortunately I'm unable to write such test:

- within MC tests, the `fixup_*` functions always witness zero for `Value`,
- within LDD tests I'm only able to trigger `relocation R_AVR_... out of range: ... is not in [..., ...]`.

I could use some more help with that.

Reply:
I suggest, separate this change of `checkUnsignedWidth` into another PR, and you only do the relocation in current PR. In this way your work in rust won't be blocked.